src/HartreeParticleDSL/Particle_IR/symbols/symboltable.py

```diff
@@ -0,0 +1,168 @@
+    
+        :raises TypeError: if the type_symbol argument is not the type of a Symbol \
+                           object class or one of its subclasses.
+        :raises TypeError: if the name argument is not a str.
+        '''
+        if not (isinstance(symbol_type, type) and
+                Symbol in inspect.getmro(symbol_type)):
+            raise TypeError(f"The symbol_type parameter should be a type class of "
+                            f"a sublcass of Symbol, but found {type(symbol_type)} instead.")
+        if not isinstance(name, str):
+            raise TypeError(f"The name parameter should be a str, but found {type(name)}.")
+    
+        visibility = Symbol.Visibility.LOCAL
+        if self._is_global:
+            visibility = Symbol.Visibility.GLOBAL
+    
+        symbol = symbol_type(name=name, datatype=datatype, visibility=visibility)
+        self.add(symbol)
+
+        return symbol
+
+    def lookup(self, name: str) -> Union[Symbol, NoneType]:
+        '''
+        Search for a Symbol named name in the symbol table. Returns the symbol
+        if it exists, otherwise returns None.
+
+        :param name: Name of the symbol to lookup.
+        :type name: str
+
+        :returns: The symbol found in the symbol table.
+        :rtype: Symbol or None.
+
+        :raises TypeError: If the name argument is not a string.
+        '''
+        if not isinstance(name, str):
+            raise TypeError("Expected the name argument to be a string, but got "
+                            f"{type(name)}.")
+
+        # TODO Should we look in the global symbol table if we don't find locally?
+        return self._symbols.get(name)            
+
+    def find_or_create(self, name: str, datatype: DataType, symbol_type: type[Symbol]) ->  Symbol:
+        '''
+        Search for a symbol in the symbol table and return it, else create it if it doesn't
+        yet exist.
+
+        :param name: Name of the symbol.
+        :type name: str
+        :param datatype: The data type of the symbol.
+        :type datatype: :py:class:`HartreeParticleDSL.Particle_IR.datatypes.datatype.DataType`
+        :param symbol_type: The Particle IR type of the symbol to create.
+        :type symbol_type: :py:type:`HartreeParticleDSL.Particle_IR.symbols.symbol.Symbol`
+
+        :raises IRGenerationError: if the symbol found doesn't match the datatype requested
+        '''
+
+        sym = self.lookup(name)
+        if sym is not None:
+            if sym.datatype != datatype:
+                raise IRGenerationError(f"Found a symbol with specified name {name}, but "
+                                        f"it had datatype {sym.datatype} which doesn't "
+                                        f"match requested datatype {datatype}.")
+            return sym
+        return self.new_symbol(name=name, datatype=datatype, symbol_type=symbol_type)
+
+
+    # Future potential requirements
+    # remove
+    # merge
```

Comment:
Merge is NYI but probably useful for merging together kernels. I think that probably doesn't make it to this PR, and if not then need an issue for it.

Reply:
Issue opened for future.